### PR TITLE
Add docker paths for MacOS

### PR DESCRIPTION
**Pull request by Matthijsy**
This PR adds the the default docker datapath for MacOS

### CODE REVIEW

**Comment by JSCU-CNI** [comment]
Could you share the folder structure you observe locally inside `$HOME/Library/Containers/com.docker.docker/Data/vms/0/data`? As I understand Docker on macOS uses a virtual machine to run containers and the `data` directory would contain e.g. `Docker.raw` or `Docker.qcow2`. I think you would have to parse that virtual disk to actually read `/var/lib/docker/*`.

**Comment by Matthijsy** [comment]
You're right, I missed that while implementing. I have a .raw on my system, not sure if it can be qcow on other systems. While implementing I found that moving that directory is not reflected in daemon.json but in `Library/Group Containers/group.com.docker/settings-store.json`.

The raw file looks like this:
```
cni
containerd
containerd-stargz-grpc
desktop-containerd
docker
dpkg
kubeadm
kubelet-plugins
lost+found
machine-id
mutagen
nfs
swap
wasm
```

The `docker` directory is comparable to `/var/lib/docker`. So after finding the installs it now mounts the docker directory such that we can use that in further processing. On my systems the repositories.json is missing, so potentially this is stored differently than on linux, but that can be investigated and fixed in a later PR.

File: dissect/target/plugins/apps/container/docker.py [hunk 1]
```diff
@@ -102,6 +103,22 @@ def __init__(self, target: Target):
...
+            if (docker_path := install.joinpath("Docker.raw")).exists():
+                vs = volume.open(container.open(docker_path))
+                if len(vs.volumes) > 1:
+                    self.target.log.warning(
+                        "Docker container %s has multiple volumes, only using first volume", install.name
+                    )
+
+                fs = filesystem.open(vs.volumes[0])
+
+                if not fs.get("/docker"):
```
Comment (Schamper) [inline dissect/target/plugins/apps/container/docker.py]: I don't think this method can return `None`? I think it raises.

In any case, preferably already use `.path()` API here. And add a unit test for the negative case.

File: dissect/target/plugins/apps/container/docker.py [hunk 2]
```diff
@@ -102,6 +103,22 @@ def __init__(self, target: Target):
...
+                    self.target.log.warning("Docker.raw misses docker directory")
+                else:
+                    self.target.fs.mount(str(install), fs.get("/docker"))
```
Comment (Schamper) [inline dissect/target/plugins/apps/container/docker.py]: I'm not a great fan of the modifications of the mounts (that would only happen when running this plugin). The source paths also won't be completely truthful.

I'm not sure if it'll actually be nicer, but a random idea I had:
- Don't mount it, but do keep a reference of the filesystem
- When yielding paths, also yield paths from the stored filesystem
- When setting source paths, somehow construct a path that resolves to `path/to/Docker.raw/path/to/container`

This works around both points, and makes it more explicit that the path comes from the `Docker.raw` file.

Let me know your thoughts!

**Review by Schamper — CHANGES_REQUESTED** [review]

**Comment by Matthijsy** [comment]
Thank you for the review @Schamper I've updated the logic to your proposed implementation. I have added tests based on two `Docker.raw` images which I generated by hand. If there is a better approach please let me know

File: dissect/target/plugins/apps/container/docker.py [hunk 3]
```diff
@@ -110,7 +111,8 @@ def check_compatible(self) -> None:
     def images(self) -> Iterator[DockerImageRecord]:
         """Returns any pulled docker images on the target system."""
         for data_root in self.installs:
-            images_path = data_root.joinpath("image/overlay2/repositories.json")
+            data_relative_root = open_docker_raw(data_root)
+            images_path = data_relative_root.joinpath("image/overlay2/repositories.json")
```
Comment (JSCU-CNI) [inline dissect/target/plugins/apps/container/docker.py]: I am not a big fan of fixing the (relative) data root paths in each exported function. This causes duplication and does not feel like the right approach. I propose to change the `find_installs` method instead.

File: dissect/target/plugins/apps/container/docker.py [hunk 4]
```diff
@@ -308,7 +324,25 @@ def get_data_path(path: Path) -> str | None:
         log.debug(exc_info=e)
         return None
 
-    return config.get("data-root")
+    return config.get("data-root") or config.get("DataFolder")
```
Comment (JSCU-CNI) [inline dissect/target/plugins/apps/container/docker.py:312]: Could you add a comment where `DataFolder` originates from (macOS-specific)?

File: dissect/target/plugins/apps/container/docker.py [hunk 5]
```diff
@@ -357,7 +403,10 @@ def find_installs(target: Target) -> Iterator[Path]:
                     target.log.info("Unable to get data-root from docker daemon file %s", config_file)
                     continue
                 if (data_root_path := target.fs.path(data_path)).exists():
-                    yield data_root_path
+                    try:
+                        yield expand_datapath_macos(data_root_path)
+                    except Exception as e:
+                        print(e)
```
Comment (JSCU-CNI) [inline dissect/target/plugins/apps/container/docker.py:392]:
```suggestion
                        target.log.warning("<a meaningful warning message here>: %s", e)
                        target.log.debug("", exc_info=e)
```

File: dissect/target/plugins/apps/container/docker.py [hunk 6]
```diff
@@ -325,6 +359,11 @@ def find_installs(target: Target) -> Iterator[Path]:
         "sysvol/ProgramData/docker",
     ]
 
+    user_data_paths = [
+        # MacOS
+        "Library/Containers/com.docker.docker/Data/vms/0/data",
```
Comment (JSCU-CNI) [inline dissect/target/plugins/apps/container/docker.py]: Technically more vms could exist if I understand the earlier referenced resources correctly. I'd opt to iterate the `vms` directory instead of assuming only `0` exists.

**Review by JSCU-CNI — CHANGES_REQUESTED** [review]

File: dissect/target/plugins/apps/container/docker.py (hunk 3 above)
Comment (Matthijsy) [reply dissect/target/plugins/apps/container/docker.py]: Would you agree than that `find_installs` can yield either a Path or a Filesystem? Than I can make it already yield the filesystem of the `Docker.raw` and just do the joinpath the same way as for a Path here.

Comment (JSCU-CNI) [reply dissect/target/plugins/apps/container/docker.py]: I don't think `find_installs` needs to return anything other than `Path`s, since we can yield a `Path` that originates from inside the `Docker.raw` container. Perhaps make `find_installs` return `tuple[Path, Path | None]` where the second `Path | None` indicates where the first `Path` originates from. That way you can reconstruct the full original path in the `source` of your records.

Comment (Matthijsy) [reply dissect/target/plugins/apps/container/docker.py]: But where would you open the container in that case? Would we still do that in every exported function? Since the full Path within the `Docker.raw` is not accessible on the target itself.

Comment (JSCU-CNI) [reply dissect/target/plugins/apps/container/docker.py]: Inside `find_installs`, something akin to:

```python
    ...
    for path in user_data_paths:
        for user_details in target.user_details.all_with_home():
            if (user_path := user_details.home_path.joinpath(path)).exists():
                if (docker_raw := user_path.joinpath("Docker.raw").is_file():
                    try:
                        vs = volume.open(container.open(docker_raw))
                        fs = filesystem.open(vs.volumes[0])
                        yield fs.path("/docker"), docker_raw
                    except ContainerError as e:
                        target.log.warning("...")
                        target.log.debug("", exc_info=e)
                else:
                    yield user_path, None
```

File: dissect/target/plugins/apps/container/docker.py (hunk 4 above)
Comment (Matthijsy) [reply dissect/target/plugins/apps/container/docker.py:312]: I cannot find any official documentation that specifies how this file should look like. I can reference the Docker Desktop settings if that adds any value? The DataFolder key is not documented there unfortunately. 
https://docs.docker.com/desktop/settings-and-maintenance/settings/
